server/src/main/java/org/opensearch/search/lookup/SearchLookup.java

```diff
@@ -153,8 +153,14 @@ public final SearchLookup forkAndTrackFieldReferences(String field) {
         return new SearchLookup(this, newFieldChain);
     }
 
+    // SourceLookup is not thread safe so we create a new instance for each leaf to support concurrent segment search
     public LeafSearchLookup getLeafSearchLookup(LeafReaderContext context) {
-        return new LeafSearchLookup(context, docMap.getLeafDocLookup(context), sourceLookup, fieldsLookup.getLeafFieldsLookup(context));
+        return new LeafSearchLookup(
+            context,
+            docMap.getLeafDocLookup(context),
+            new SourceLookup(),
```

Comment:
Overall looks fine to me, the only caveat is the `SearchLookup` returns this `SourceLookup` which it creates to the caller. So if there is some state which is expected to be set in the `SourceLookup` of this `SearchLookup` via the leafSearchLookup mechanism that won't be the case anymore. I looked into one usage in fetch context and seems like there it again creates the new `SearchLookup` and uses its `SourceLookup` to fill the docs, so that should be fine. Can you double check on the consumers of the `source`. Should we also add comment to the `source` method that the `SourceLookup` returned will not be correlated to `LeafSearchLookup` in anyway ?

Reply:
I also saw the same that it is only used in fetch context. Will add a comment with the warning.

Reply:
It seems like HitContext (the only consumer of `SearchLookup::lookup()`) should be fine, it does call `SourceLookup::setSegmentAndDocument`

Reply:
I might be overthinking, but since `source()` is a public API, so plugins might be making use of it. So if they rely on the fact the `sourceLookup` is only created once and use this instance to get the information about current segment being operated or make use of some other state, that may no more hold true. So should we also set `this. sourceLookup` here?  such that when `source()` is called, they get the latest/last `sourceLookup`.

Reply:
> So should we also set this. sourceLookup here? such that when source() is called, they get the latest/last sourceLookup.

I think we should not change `this.sourceLookup` here

Reply:
I'm in favour of deprecation but deprecated APIs should break. 
`searchLookup.source().loadSourceIfNeeded()` seems like a common pattern of usage on searching on github. So this will break since it will keep referring to the old `sourceLookup`.

Reply:
@rishabhmaurya could you share a GitHub search or some references to where `searchLookup.source().loadSourceIfNeeded()` is used? From what I can see the usage is `LeafSearchLookup#source` which gets the `SourceLookup` associated with the `LeafSearchLookup`.

Generally speaking I don't think `searchLookup.source().loadSourceIfNeeded()` is a safe usage because if `setSegmentAndDocument` is not first called then your `fieldReader` won't be set and it looks like you'd get an exception.

Reply:
https://github.com/search?q=SearchLookup+source%28%29&type=code 
most of these references are pointing to people forking off elasticsearch, so i didn't open the links to check the exact use cases.

Reply:
> Generally speaking I don't think searchLookup.source().loadSourceIfNeeded() is a safe usage because if setSegmentAndDocument is not first called then your fieldReader won't be set and it looks like you'd get an exception.

But what if `setSegmentAndDocument` is called and then they use `searchLookup.source().loadSourceIfNeeded()`

Reply:
So I think what you're describing is if someone does this:
1. `leafLookup = SearchLookup.getLeafSearchLookup`
2. `leafLookup.setDocument`
3. `SearchLookup.source.loadSourceIfNeeded`

Then this is broken by the changes in this PR. I don't think this is a valid use case though because in step 3 users should be doing `leafLookup.source.loadSourceIfNeeded` to guarantee they're addressing the right segment.

I still think the changes in this PR are fine but I'm also fine with setting `this.sourceLookup` like you originally suggested but that still wouldn't work for the concurrent search case as multiple segments get operated on at the same time so this sourceLookup instance may not be referring to the latest source lookup.

Reply:
Actually I think in order to set `this.sourceLookup` in `SearchLookup` we would have to make `getLeafSearchLookup` synchronized because we're still calling this `getLeafSearchLookup` across multiple threads at the same time.

> Usually, when people don't like passing leafLookup() across methods down in the chain, but they know they have QueryShardContext available, from which they can get SearchLookup, I can see above pattern getting abused.

IMO this is more of a bug/invalid usage and I don't think we need to keep the backwards compatibility for this especially as we are now documenting the correct way to use it in the code comments. Let me know if you think this PR is good to merge @rishabhmaurya

Reply:
sure, let's go ahead. Thanks @jed326